microprofile/WFLY-15832-microconfig-config-root-folder.adoc

```diff
@@ -0,0 +1,99 @@
+
+== Issue Metadata
+
+=== Issue
+
+https://issues.redhat.com/browse/WFLY-15832[WFLY-15832 - MP Config: Ability to specify root config source directory]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1783[EAP7-1783 - MP Config: Ability to specify root config source directory]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE. 
+// Discuss with QE during the Kickoff state to decide this
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+All code changes are in the MicroProfile Config with SmallRye subsystem in WildFly.
+
+=== Other Interested Projects
+N/A
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+* The `dir` complex attribute of the `/subsystem=microprofile-config-smallrye/config-source=*` resource will be extended to have an additional `root` nested attribute.
+** The default value of the `root` nested attribute is `false`, giving the legacy behaviour
+** If `root=true`, the directory is treated as a root directory for ConfigSources
+* Within a root directory for ConfigSource:
+** Files contained directly within the root directory are ignored and are not used for config
+** Each top-level directory is treated as a ConfigSource
+** Files within each top-level directory under the root directory are treated as a config entry
+** Only top-level directories are considered, directories nested within those are ignored
+** By default the ConfigSources for each top level directory get the `ordinal` of the `/subsystem=microprofile-config-smallrye/config-source=*` resource used to add the root directory
+*** If the top-level directory contains a `config_ordinal` file, the value specified in there will override the default `ordinal`
+** If two top-level directories with the same `ordinal` contain the same entry, their names are sorted according to standard Java sorting rules and the first one is used
+
+=== Nice-to-Have Requirements
+N/A
+
+=== Non-Requirements
+* As mentioned previously, in <<hard-requirements, Hard Requirements>>, only top level directories under a ConfigSource root directory will be considered as ConfigSources. We don't check nested folders.
+* No attempts will be made to avoid having multiple ConfigSources defined in the subsystem referencing the same directories
+
+== Backwards Compatibility
+The feature is backward compatible.
+
+== Security Considerations
+None. The feature doesn't expose anything which would not be possible with what is there today.
```

Comment:
Is it true? So far EAP administrator had to configure each file explicitely. So administrator guarantees files are correct. With this feature in place, attacker (e.g. other user on OpenShift) can prepare new (unchecked by administrator) files to be processed by EAP. E.g. by putting something into `/bindings` or `/etc/config`. Those files could be malformed (some sort of injection) or a lot of them could be there (DoS attack). Or it can be file which "tricks" ordinal resolving https://github.com/wildfly/wildfly-proposals/pull/450/files#diff-9f1354c78479d2107750d078d4f846420ac82216e6eda99d34a0ad36ea2d948fR77 and value expected by administrator can be overwritten by attacker.

Reply:
Does not need to be attacker. It can happen easily by accident. If `/bindings` and `/etc/config` are common places where OpenShift store ServiceBindings and ConfigMaps from whole cluster. It can happen easily (e.g. on managed cluster like Dedicated or Sandbox instances) key collision will happen and which value will be used will be determined by ordinal resolving (alphabetical order?) https://github.com/wildfly/wildfly-proposals/pull/450/files#diff-9f1354c78479d2107750d078d4f846420ac82216e6eda99d34a0ad36ea2d948fR77. This feature seems very dangerous to me.

Reply:
> attacker (e.g. other user on OpenShift)

The ConfigMap must be in the same namespace than the application. In essence, if an OpenShift user can do that, he also has direct access to the application itself.

Reply:
your 2nd point is correct but that would be a misconfiguration, not a security issue.

We could have 2 service bindings injection config map that would provide the same `USERNAME` property for example.
However there are mechanism to avoid this and have deterministic configuration such as having unambiguous prefix for different service bindings.

Reply:
From wording of [1] and [2] I understood `/bindings` and `/etc/config` we are talking about are OpenShift wide directories. So something from different namespace will be visible in EAP pod. So you are basically saying on OpenShift level it is guaranteed in EAP pod in `/bindings` and `/etc/config` there is mounted only stuff related to Pods namespace? If this is true my concern is not valid. And that would make sense.

But in this case could you make it please in [1], [2] explicit. So any other reader have a chance to misundarstand it like I did. 

[1] https://github.com/wildfly/wildfly-proposals/pull/450/files#diff-9f1354c78479d2107750d078d4f846420ac82216e6eda99d34a0ad36ea2d948fR14
[2] https://github.com/wildfly/wildfly-proposals/pull/450/files#diff-9f1354c78479d2107750d078d4f846420ac82216e6eda99d34a0ad36ea2d948fR16

Reply:
https://docs.openshift.com/container-platform/4.10/nodes/pods/nodes-pods-configmaps.html

> ConfigMap objects reside in a project.
> They can only be referenced by pods in the same project.

Reply:
I just checked some EAP pod and I really don't see any `/bindings` (I don't have ServiceBinding) and `/etc/config` (I have ConfigMap in namespace) directories available automatically. So I assume someone needs to configure that. 

Especially I am questioning this part how is meant. 

```
ConfigMaps are installed under `/etc/config`.

So if you have two ConfigMaps called `cfg-a` and `cfg-b`, their respective directories containing the config files will be `/etc/config/cfg-a` and `/etc/config/cfg-b`.
```